spi/src/main/java/org/jboss/pnc/spi/datastore/predicates/BuildRecordPredicates.java

```diff
@@ -124,6 +125,139 @@
         };
     }
 
+    /**
+     * (related to NCL-6957)
+     *
+     * When (re)building a temporary build: include either persistent and temporary builds
+     * 
+     * When (re)building a persistent build: include only persistent builds
+     * 
+     * @param temporary if requested (re)build is temporary
+     * @return Predicate that filters out builds according to description
+     */
+    public static Predicate<BuildRecord> includeTemporary(boolean temporary) {
+        return (root, query, cb) -> (temporary) ? cb.and() : cb.isFalse(root.get(BuildRecord_.temporaryBuild));
+    }
+
+    /**
+     * Changed behavior after NCL-6895 Dry-Run implementation
+     *
+     * When (re)building a persistent build: - include only existing persistent builds having the same idRev
+     *
+     * When (re)building a temporary PREFER_PERSISTENT build: if there are existing successful persistent builds having
+     * the same idRev, ignore temporary builds, otherwise include also successful temporary builds having the same idRev
+     * 
+     * When (re)building a temporary PREFER_TEMPORARY build: if there are existing successful temporary builds having
+     * the same idRev, ignore persistent builds, otherwise include also successful persistent builds having the same
+     * idRev
+     *
+     * @param idRev the revision of the build to (re)build
+     * @param temporary if requested (re)build is temporary
+     * @param alignmentPreference the alignmentPreference specified in case of temporary builds
+     *
+     * @return Predicate that filters out builds according to description
+     */
+    public static Predicate<BuildRecord> includeTemporary(
+            IdRev idRev,
+            boolean temporary,
+            AlignmentPreference alignmentPreference) {
+        return (root, query, cb) -> {
+            if (temporary) {
+                if (AlignmentPreference.PREFER_TEMPORARY.equals(alignmentPreference)) {
+                    Subquery<Long> temporaryCount = query.subquery(Long.class);
+                    Root<BuildRecord> subRoot = temporaryCount.from(BuildRecord.class);
+                    temporaryCount.select(cb.count(subRoot.get(BuildRecord_.id)));
+                    temporaryCount.where(
+                            cb.and(
+                                    cb.isTrue(subRoot.get(BuildRecord_.temporaryBuild)),
+                                    cb.equal(subRoot.get(BuildRecord_.status), BuildStatus.SUCCESS),
+                                    cb.equal(subRoot.get(BuildRecord_.buildConfigurationId), idRev.getId()),
+                                    cb.equal(subRoot.get(BuildRecord_.buildConfigurationRev), idRev.getRev())));
```

Comment:
```suggestion                    
```
This subquery should actually look at all builds in the history of the BC and get a count of temporary builds. Looking at just the ID REV is not what DA does as DA doesn't know about IDREVs and just looks at what versions have been built. So maybe just the method `includeTemporary( Integer configurationId, boolean temporary, AlignmentPreference alignmentPreference)` is required and this one can actually be deleted. 

The logic for figuring out whether a dependency 'has newer version' should be the same for all dependencies, doesn't matter if they are explicit or implicit.

Reply:
I have made the change to use the same method (checking only for the same build config) .... and I left all the previous methods on purpose in case we need to switch back :-)